pkg/operator/controller/controller_router_deployment.go

```diff
@@ -0,0 +1,245 @@
+// clusteringress.
+func (r *reconciler) ensureRouterDeployment(ci *ingressv1alpha1.ClusterIngress, ingressDomain string, infraConfig *configv1.Infrastructure, ha ingressv1alpha1.ClusterIngressHighAvailability) (*appsv1.Deployment, error) {
+	desired, err := desiredRouterDeployment(ci, ingressDomain, r.Config.RouterImage, infraConfig, ha.Type)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build router deployment: %v", err)
+	}
+	current, err := r.currentRouterDeployment(ci)
+	if err != nil {
+		return nil, err
+	}
+	switch {
+	case desired != nil && current == nil:
+		if err := r.createRouterDeployment(desired); err != nil {
+			return nil, err
+		}
+	case desired != nil && current != nil:
+		if err := r.updateRouterDeployment(current, desired); err != nil {
+			return nil, err
+		}
+	}
+	return r.currentRouterDeployment(ci)
+}
+
+// ensureRouterDeleted ensures that any router resources associated with the
+// clusteringress are deleted.
+func (r *reconciler) ensureRouterDeleted(ci *ingressv1alpha1.ClusterIngress) error {
+	deployment := manifests.RouterDeployment(ci)
+	name := routerDeploymentName(ci)
+	deployment.Name = name.Name
+	deployment.Namespace = name.Namespace
+	err := r.Client.Delete(context.TODO(), deployment)
+	if !errors.IsNotFound(err) {
+		return err
+	}
+	return nil
+}
+
+// desiredRouterDeployment returns the desired router deployment.
+func desiredRouterDeployment(ci *ingressv1alpha1.ClusterIngress, ingressDomain string, routerImage string, infraConfig *configv1.Infrastructure, haType ingressv1alpha1.ClusterIngressHAType) (*appsv1.Deployment, error) {
+	deployment := manifests.RouterDeployment(ci)
+	name := routerDeploymentName(ci)
+	deployment.Name = name.Name
+	deployment.Namespace = name.Namespace
+
+	if deployment.Labels == nil {
+		deployment.Labels = map[string]string{}
+	}
+	deployment.Labels[manifests.OwningClusterIngressLabel] = ci.Name
+
+	if deployment.Spec.Template.Labels == nil {
+		deployment.Spec.Template.Labels = map[string]string{}
+	}
+	deployment.Spec.Template.Labels["router"] = deployment.Name
+
+	if deployment.Spec.Selector.MatchLabels == nil {
+		deployment.Spec.Selector.MatchLabels = map[string]string{}
+	}
+	deployment.Spec.Selector.MatchLabels["router"] = deployment.Name
+
+	statsSecretName := fmt.Sprintf("router-stats-%s", ci.Name)
+	env := []corev1.EnvVar{
+		{Name: "ROUTER_SERVICE_NAME", Value: ci.Name},
+		{Name: "STATS_USERNAME", ValueFrom: &corev1.EnvVarSource{
+			SecretKeyRef: &corev1.SecretKeySelector{
+				LocalObjectReference: corev1.LocalObjectReference{
+					Name: statsSecretName,
+				},
+				Key: "statsUsername",
+			},
+		}},
+		{Name: "STATS_PASSWORD", ValueFrom: &corev1.EnvVarSource{
+			SecretKeyRef: &corev1.SecretKeySelector{
+				LocalObjectReference: corev1.LocalObjectReference{
+					Name: statsSecretName,
+				},
+				Key: "statsPassword",
+			},
+		}},
+	}
+
+	// Enable prometheus metrics
+	certsSecretName := fmt.Sprintf("router-metrics-certs-%s", ci.Name)
```

Comment:
Just a note for followups: isn't this a hidden dependency on `manifests.RouterServiceInternal`?

Reply:
Yeah, there's more refactoring to be done around secrets' names.